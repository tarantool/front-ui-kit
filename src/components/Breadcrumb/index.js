// @flow
import * as React from 'react';
import { css, cx } from 'emotion';
import { Text } from '../Text';
import { OverflowList } from '../OverflowList';
import { DropdownItem, Dropdown } from '../Dropdown';

import BreadcrumbItemComponent from './components/BreadcrumbItem';

export type BreadcrumbItem = {
  title: string,
  link?: string,
}

export type ActionsBreadCrumbs = {
  onLinkClick?: (link: string) => void
}

type BreadcrumbProps = ActionsBreadCrumbs & {
  breadcrumbs: BreadcrumbItem[],
  appName?: string,
}

const styles = {
  breadcrumbs: css`
    display: flex;
  `,
  breadcrumbElement: css`
    white-space: nowrap;
  `,
  breadcrumbsList: css`
    width: 100%;
  `,
  breadcrumbDelimeter: css`
    margin: 0 6px;
    color: #686D77;
    font-size: 16px;
    line-height: 22px;
    white-space: nowrap;
  `,
  appName: css`
    position: absolute;
    white-space: nowrap;
    left: 0;
    top: 0;
    background: #FFFFFF;
    box-shadow: 0px 0px 10px #E6E7E8;
    border-radius: 2px;
    margin-left: -10px;
    padding: 0 10px;
  `,
  breadcrumbHoveredElement: css`
    position: relative;
    cursor: pointer;
  `,
  overflowButton: css`
    cursor: pointer;
    font-size: 16px;
    line-height: 22px;
    color: #8E9199;
    
    &:hover {
      text-decoration-line: underline;
    }
  `
};

const MAX_APP_NAME_LENGTH = 40;

export class Breadcrumb extends React.Component<BreadcrumbProps>{
  render() {
    const { breadcrumbs, appName } = this.props;
    const isLongAppName = (appName: string) => appName.length > MAX_APP_NAME_LENGTH;
    return (
      <div className={styles.breadcrumbs}>
        {appName &&
        (isLongAppName(appName)
          ? <AppName appName={appName} />
          :
          <Text
            tag="span"
            className={styles.breadcrumbElement}
            variant="h3"
          >
            {appName}
          </Text>)
        }
        <OverflowList
          minVisibleItems={1}
          items={breadcrumbs}
          className={styles.breadcrumbsList}
          overflowRenderer={this.renderOverflow}
          visibleItemRenderer={this.renderBreadcrumbWrapper}
        />
      </div>
    );
  }

  renderOverflow = (items: any[]) => {
    const { onLinkClick } = this.props;
    const itemsCollection = (
        <>
          {items.map(item => (
            <DropdownItem onClick={() => onLinkClick && onLinkClick(item.link)}>
              <Text className={styles.overflowButton}>{item.title}</Text>
            </DropdownItem>
          ))}
        </>
    );

    return (
      <React.Fragment>
        <Text tag="span" className={cx(styles.breadcrumbDelimeter, styles.breadcrumbElement)}>
          /
        </Text>
        <Dropdown items={itemsCollection}>
          <Text className={styles.overflowButton}>...</Text>
        </Dropdown>
      </React.Fragment>
    )
  };

  renderBreadcrumbWrapper = (props: BreadcrumbItem, index: number) => {
    const { onLinkClick } = this.props;
    return (
      <React.Fragment>
        <Text tag="span" className={cx(styles.breadcrumbDelimeter, styles.breadcrumbElement)}>
          /
        </Text>
        <BreadcrumbItemComponent key={index} title={props.title} link={props.link} onLinkClick={onLinkClick} />
      </React.Fragment>
    )
  };
}


function AppName({ appName }) {
  const [ isHover, setHover ] = React.useState(false);
  const shoterAppName = (appName: string) => appName.length > MAX_APP_NAME_LENGTH
    ? `${appName.slice(0, MAX_APP_NAME_LENGTH)}...`
    : appName;

  return (
    <Text
      tag="span"
      className={cx(styles.breadcrumbElement, styles.breadcrumbHoveredElement)}
      variant="h3"
      onMouseOver={() => setHover(true)}
      onMouseLeave={() => setHover(false)}
    >
      {shoterAppName(appName)}
      {isHover &&
        <Text
          tag="span"
          className={cx(styles.breadcrumbElement, styles.appName)}
          variant="h3"
        >
          {appName}
        </Text>
      }
    </Text>
  )
}
