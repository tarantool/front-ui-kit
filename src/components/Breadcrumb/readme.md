```js
const breadcrumbs = [{
  title: 'Settings',
},
{
  title: 'Roles',
},
{
  title: 'Roles',
},
{
  title: 'Roles',
},
{
  title: 'Roles',
},
{
  title: 'Roles',
},
{
  title: 'Roles',
},
{
  title: 'Roles',
},
{
  title: 'Roles',
},
{
  title: 'Roles',
},
{
  title: 'Roles',
},
{
  title: 'Roles',
},
{
  title: 'Roles',
},
{
  title: 'Roles',
},
{
  title: 'Roles',
},
{
  title: 'Create role',
}];
<div>
<Breadcrumb
 breadcrumbs={breadcrumbs}
 appName={'Project name-2948474770249-999-20299547789'}
 />

<Breadcrumb
 breadcrumbs={breadcrumbs}
 appName={'Project name'}
 />
</div>


```
