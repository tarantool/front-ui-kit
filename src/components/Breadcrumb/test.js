import React from 'react';
import renderer from 'react-test-renderer';
import { Breadcrumb } from '../Breadcrumb';

describe('Breadcrumb', () => {
  const breadcrumbs = [{
    title: '@tarantool.io/ui-kit',
    link: '/tarantool'
  },
  {
    title: 'UI Components',
    link: '/section-ui-components'
  },
  {
    title: 'Breadcrumb',
    link: '/breadcrumb'
  },
  {
    title: 'OverflowList',
    link: '/overflow-list'
  },
  {
    title: 'ResizeSensor',
    link: '/resize-sensor'
  }];
  const onLinkClick = () => {};

  it('renders correctly', () => {
    const tree = renderer.create(
      <>
        <Breadcrumb
          breadcrumbs={breadcrumbs}
          appName={'Project name'}
          onLinkClick={onLinkClick}
        />
      </>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });


  it('renders correctly long name', () => {
    const tree = renderer.create(
      <>
        <Breadcrumb
          breadcrumbs={breadcrumbs}
          appName={'Project name-2948474770249-999-20299547789123123141'}
          onLinkClick={onLinkClick}
        />
      </>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('block with small width', () => {
    const tree = renderer.create(
      <>
        <div style={{ width: '300px' }}>
          <Breadcrumb
            breadcrumbs={breadcrumbs}
            appName={'Project name-2948474770249-999-20299547789123123141'}
            onLinkClick={onLinkClick}
          />
        </div>
      </>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
