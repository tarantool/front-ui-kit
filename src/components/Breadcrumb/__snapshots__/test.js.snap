// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumb block with small width 1`] = `
<div
  style={
    Object {
      "width": "300px",
    }
  }
>
  <div
    className="css-k008qs"
  >
    <span
      className="css-1u31squ"
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
    >
      Project name-2948474770249-999-202995477...
    </span>
    <div
      className="css-vzwfet"
    >
      <span
        className="css-1tb5g3v"
      >
        /
      </span>
      <a
        className="css-g4sqw9"
        href="/tarantool"
        onClick={[Function]}
      >
        @tarantool.io/ui-kit
      </a>
      <span
        className="css-1tb5g3v"
      >
        /
      </span>
      <a
        className="css-g4sqw9"
        href="/section-ui-components"
        onClick={[Function]}
      >
        UI Components
      </a>
      <span
        className="css-1tb5g3v"
      >
        /
      </span>
      <a
        className="css-g4sqw9"
        href="/breadcrumb"
        onClick={[Function]}
      >
        Breadcrumb
      </a>
      <span
        className="css-1tb5g3v"
      >
        /
      </span>
      <a
        className="css-g4sqw9"
        href="/overflow-list"
        onClick={[Function]}
      >
        OverflowList
      </a>
      <span
        className="css-1tb5g3v"
      >
        /
      </span>
      <a
        className="css-g4sqw9"
        href="/resize-sensor"
        onClick={[Function]}
      >
        ResizeSensor
      </a>
      <div
        className="css-11rp940"
      />
    </div>
  </div>
</div>
`;

exports[`Breadcrumb renders correctly 1`] = `
<div
  className="css-k008qs"
>
  <span
    className="css-4haz1u"
  >
    Project name
  </span>
  <div
    className="css-vzwfet"
  >
    <span
      className="css-1tb5g3v"
    >
      /
    </span>
    <a
      className="css-g4sqw9"
      href="/tarantool"
      onClick={[Function]}
    >
      @tarantool.io/ui-kit
    </a>
    <span
      className="css-1tb5g3v"
    >
      /
    </span>
    <a
      className="css-g4sqw9"
      href="/section-ui-components"
      onClick={[Function]}
    >
      UI Components
    </a>
    <span
      className="css-1tb5g3v"
    >
      /
    </span>
    <a
      className="css-g4sqw9"
      href="/breadcrumb"
      onClick={[Function]}
    >
      Breadcrumb
    </a>
    <span
      className="css-1tb5g3v"
    >
      /
    </span>
    <a
      className="css-g4sqw9"
      href="/overflow-list"
      onClick={[Function]}
    >
      OverflowList
    </a>
    <span
      className="css-1tb5g3v"
    >
      /
    </span>
    <a
      className="css-g4sqw9"
      href="/resize-sensor"
      onClick={[Function]}
    >
      ResizeSensor
    </a>
    <div
      className="css-11rp940"
    />
  </div>
</div>
`;

exports[`Breadcrumb renders correctly long name 1`] = `
<div
  className="css-k008qs"
>
  <span
    className="css-1u31squ"
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
  >
    Project name-2948474770249-999-202995477...
  </span>
  <div
    className="css-vzwfet"
  >
    <span
      className="css-1tb5g3v"
    >
      /
    </span>
    <a
      className="css-g4sqw9"
      href="/tarantool"
      onClick={[Function]}
    >
      @tarantool.io/ui-kit
    </a>
    <span
      className="css-1tb5g3v"
    >
      /
    </span>
    <a
      className="css-g4sqw9"
      href="/section-ui-components"
      onClick={[Function]}
    >
      UI Components
    </a>
    <span
      className="css-1tb5g3v"
    >
      /
    </span>
    <a
      className="css-g4sqw9"
      href="/breadcrumb"
      onClick={[Function]}
    >
      Breadcrumb
    </a>
    <span
      className="css-1tb5g3v"
    >
      /
    </span>
    <a
      className="css-g4sqw9"
      href="/overflow-list"
      onClick={[Function]}
    >
      OverflowList
    </a>
    <span
      className="css-1tb5g3v"
    >
      /
    </span>
    <a
      className="css-g4sqw9"
      href="/resize-sensor"
      onClick={[Function]}
    >
      ResizeSensor
    </a>
    <div
      className="css-11rp940"
    />
  </div>
</div>
`;
